File: Go/proj_euler_problem_0004.go
/*
 * Problem 4
 * Largest palindrome product
 *
 * A palindromic number reads the same both ways. The largest palindrome
 * made from the product of two 2-digit numbers is 9009 = 91 × 99.
 * 
 * Find the largest palindrome made from the product of two 3-digit numbers.
 *
 */

package main

import (
    "fmt"
    "strconv"
)

func isPalindrome(a string) bool {
    var length int = len(a)
    for i := 0; i < length / 2; i++ {
        if a[i] != a[length - i - 1] {
            return false
        }
    }
    return true
}

func main() {
    var a int = 100
    var b int = 100
    var ab int = 1
    var ab_s string = ""
    var max_palindrome = 1
    
    for a = 999; a >= 100; a-- {
        if a * 999 < max_palindrome {
            break
        }
        for b = 999; b >= 100; b-- {
            ab = a * b
            if ab < max_palindrome {
                break
            }
            ab_s = strconv.FormatInt(int64(ab), 10)
            if isPalindrome(ab_s) {
                if ab > max_palindrome {
                    max_palindrome = ab
                }
            }
        }
    }
    
    fmt.Println(max_palindrome)
}    

File: Python/proj_euler_problem_0004.py
'''
Problem 4
Largest palindrome product

A palindromic number reads the same both ways. The largest palindrome
made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.
'''

import math
import sys

palindromes = []
def isPalindrome(number):
	n = number
	reverse = 0
	while(n > 0):
		reverse = reverse * 10 + n % 10
		n = math.floor(n / 10)
	return number == reverse

def EulerProblem0004():
	for i in range(100, 1000, 1):
		for j in range(100, 1000, 1):
			if isPalindrome(i * j):
				palindromes.append(i * j)

	maxPalindrome = max(palindromes)

	print(maxPalindrome)
if __name__ == "__main__":
	EulerProblem0004()
	
